src/integration/java/io/pinecone/integration/dataplane/UpsertAndDeleteTest.java

```diff
@@ -0,0 +1,240 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class UpsertAndDeleteTest {
+    private static VectorServiceGrpc.VectorServiceBlockingStub blockingStub;
+    private static VectorServiceGrpc.VectorServiceFutureStub futureStub;
+    private static final int dimension = 3;
+
+    @BeforeAll
+    public static void setUp() throws IOException, InterruptedException {
+        PineconeConnection connection = new IndexManager().createIndexIfNotExists(dimension);
+        blockingStub = connection.getBlockingStub();
+        futureStub = connection.getFutureStub();
+    }
+
+    @Test
+    public void UpsertVectorsAndDeleteByIdSyncTest() throws InterruptedException {
+        // Upsert vectors with required parameters
+        String namespace = RandomStringBuilder.build("ns", 8);
+        List<String> upsertIds = Arrays.asList("v1", "v2", "v3");
+        UpsertResponse upsertResponse = blockingStub.upsert(buildRequiredUpsertRequest(upsertIds, namespace));
+
+        // Get vector count before deleting vectors
+        DescribeIndexStatsRequest describeIndexRequest = DescribeIndexStatsRequest.newBuilder().build();
+        DescribeIndexStatsResponse describeIndexStatsResponse = blockingStub.describeIndexStats(describeIndexRequest);
+        int startVectorCount = describeIndexStatsResponse.getNamespacesMap().get(namespace).getVectorCount();
+        assertEquals(startVectorCount, upsertResponse.getUpsertedCount());
+
+        // Delete 1 vector
+        String[] idsToDelete = new String[]{upsertIds.get(0)};
+        DeleteRequest deleteRequest = DeleteRequest.newBuilder()
+                .setNamespace(namespace)
+                .addAllIds(Arrays.asList(idsToDelete))
+                .setDeleteAll(false)
+                .build();
+        blockingStub.delete(deleteRequest);
+        Thread.sleep(3500);
+
+        // Call describeIndexStats to get updated counts
+        describeIndexStatsResponse = blockingStub.describeIndexStats(describeIndexRequest);
+        // Verify updated vector count
+        assertEquals(describeIndexStatsResponse.getNamespacesMap().get(namespace).getVectorCount(), startVectorCount - idsToDelete.length);
+
+        // Delete multiple vectors
+        idsToDelete = new String[]{upsertIds.get(0), upsertIds.get(2)};
+        deleteRequest = DeleteRequest.newBuilder()
+                .setNamespace(namespace)
+                .addAllIds(Arrays.asList(idsToDelete))
+                .setDeleteAll(false)
+                .build();
+        blockingStub.delete(deleteRequest);
+        Thread.sleep(3500);
+
+        // Call describeIndexStats to get updated counts
+        describeIndexStatsResponse = blockingStub.describeIndexStats(describeIndexRequest);
+        // Verify the updated vector count
+        assertEquals(describeIndexStatsResponse.getNamespacesMap().get(namespace).getVectorCount(), startVectorCount - idsToDelete.length);
```

Comment:
Does the `startVectorCount - idsToDelete.length` actually work here? Up above on ln.48 we're deleting `upsertIds.get(0)`. The `startVectorCount - idsToDelete.length` count works in this case as we've deleted 1 of 3 vectors.

Here though we're resetting `idsToDelete` and reusing `upsertIds.get(0)` on ln.63. So while `startVectorCount - idsToDelete.length` still validates, we're really only deleting 1 additional vector since we already deleted `upsertIds[0]`.

I was expecting ln.63 to be `idsToDelete = new String[]{upsertIds.get(1), upsertIds.get(2)};`, in which case `startVectorCount - idsToDelete.length` would equal 1, but we'd deleted 3 vectors. I'm wondering if it makes sense to hard-code things to keep legibility of the tests a bit clearer, but maybe I'm over thinking.

Reply:
You are right, the reason this test was working is because I was using `idsToDelete = new String[]{upsertIds.get(1), upsertIds.get(2)}` instead of `idsToDelete = new String[]{upsertIds.get(1), upsertIds.get(2)}`. Good catch!!